metrics.py

```diff
@@ -0,0 +1,185 @@
+# IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+# ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+# LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+# CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+# SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+# CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+# ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+# POSSIBILITY OF SUCH DAMAGE.
+"""Used to collect anonymous DeepVariant metrics."""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import atexit
+import functools
+import json
+import logging
+import time
+import uuid
+
+import requests
+from typing import Dict, Optional
+
+_CLEARCUT_ENDPOINT = 'https://play.googleapis.com/log'
+_CLOUD_HCLS = 'CLOUD_HCLS'
+_CONCORD = 'CONCORD'
+_DEEP_VARIANT_RUN = 'DeepVariantRun'
+_HTTP_REQUEST_TIMEOUT_SEC = 10
+_PYTHON = 'PYTHON'
+_VIRTUAL_CHC_DEEPVARIANT = 'virtual.chc.deepvariant'
+
+
+def capture_exceptions(func):
+  """Function decorator to capture and log any exceptions."""
+
+  @functools.wraps(func)
+  def wrapper(*args, **kwds):
+    try:
+      return func(*args, **kwds)
+    # pylint:disable=broad-except
+    except Exception as e:
+      logging.error('Exception captured in %s : %s', func.__name__, e)
+
+  return wrapper
+
+
+class _ConcordEvent(object):
+  """Encapsulates information representing a Concord event."""
+
+  def __init__(self,
+               event_name: str,
+               event_type: str,
+               project_number: int,
+               console_type: str,
+               page_hostname: str,
+               event_metadata: Optional[Dict[str, str]] = None) -> None:
+    self._event_name = event_name
+    self._event_type = event_type
+    self._project_number = project_number
+    self._console_type = console_type
+    self._page_hostname = page_hostname
+    self._event_metadata = event_metadata or {}
+
+  def to_json(self, **kwargs):
+    """Encodes data in json."""
+    event_dict = {
+        'project_number': self._project_number,
+        'event_name': self._event_name,
+        'event_type': self._event_type,
+        'console_type': self._console_type,
+        'page_hostname': self._page_hostname,
+        'event_metadata': self._event_metadata_as_kv(),
+    }
+    return json.dumps(event_dict, **kwargs)
+
+  def _event_metadata_as_kv(self):
+    kv_list = []
+    for k, v in sorted(self._event_metadata.items()):
+      kv_list.append({'key': k, 'value': str(v)})
+
+    return kv_list
+
+
+class _MetricsCollector(object):
+  """A class that collects and submits metrics.
+
+  Instances of this class share the same internal state, and thus behave the
+  same all the time.
+  """
+  _shared_events = []
```

Comment:
I think the `shared` is confusing. It reminds me that it would be accessed by multiple threads but which is not the case. It is a collector that collecting multiple events from different places. I think  `_events` is good enough. Similarly, I would prefer `_session_identifier`.  But it is up to you. If you decide to change, the above comments `Instances of this class share the same internal state, and thus behave the
  same all the time.` can also be removed. :)